### PR TITLE
doc: add guideline for RTVM workload design

### PR DESCRIPTION
**Pull request by JasonChenCJ**
a quick guideline document help RTOS designer/developer to design
his/her RT workload based on RTVM

Signed-off-by: Jason Chen CJ <jason.cj.chen@intel.com>

### CODE REVIEW

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 1]
```diff
@@ -0,0 +1,42 @@
+.. _rtvm_workload_guideline:
+
+Guidelines to the workload design in an RTVM
+############################################
+
+For a RTOS designer/developer, if he/she want to run the RTOS based on ARCN RTVM,
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: It's best to avoid using ``he/she``, so instead, write this as:
```
An RTOS developer must be aware of the differences between
running applications on a native platform and a real time virtual
machine (RTVM), especially issues impacting real time
performance.  For example, a real time thread should
avoid using VM-Exit.  If VM-Exit is used, the developer must
account for an additional margin of CPU cycles for
the added run time cost.

This document provides some workload design guidelines when
using an RTVM within the ACRN hypervisor.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+he/she must be aware of the difference between native platform and virtual machine,
+especially those will impact real time capability.
+Generally speaking, the designer/developer should try to avoid VM-Exit when running
+a RT thread, or he/she must consider adding additional margin of CPU cycles for the
+cost of VM-Exit.
+
+Best Practice: Run RTVM with dedicated resources/devices
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: Is this the only topic that's a "Best Practice"? I'd recommend leaving off the "Best Practice:" beginning of this title.

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+********************************************************
+
+ACRN allocates dedicated CPU, memory resources, cache resources (with Cache Allocations
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
ACRN allocates dedicated CPU, memory resources, and cache resources (using  Cache Allocations
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+Technology, i.e. CAT)  to RTVMs. For I/O devices, we suggest to use dedicated (pass-thru)
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
Technology, i.e. CAT)  to RTVMs. For best real time performance  of
I/O devices, we recommend using dedicated (pass-thru)
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+PCIe devices to avoid VM-Exit at run time for best real time performance.
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
PCIe devices to avoid VM-Exit at run time.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+NOTE: for pass-thru PCI device, its configure space is still emulated and the access will
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
NOTE: The configuration space for pass-thru PCI devices is still emulated and the access will
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+trigger VM-Exit.
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
trigger a VM-Exit.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+RTVM with virtio PMD(Polling Mode Driver) for I/O sharing
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
RTVM with virtio PMD (Polling Mode Driver) for I/O sharing
```
(and lengthen the heading underline)

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+*********************************************************
+
+If the RTVM has to use shared devices for some reason, ACRN recommend to use PMD drivers,
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: How about:
```
If the RTVM must use shared devices, we recommend using PMD drivers
that can eliminate the unpredictable latency caused by guest I/O
trap-and-emulate access. The RTVM workload must be aware that the
packets in the PMD driver may arrive or be sent later than expected. 
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+which can eliminate the unpredictable latency in trap-and-emulate the guest I/O access.
+However, one needs to be aware that the packet in PMD driver may arrive or be sent later
+than expected. The workload in RT VM must be aware of for better RT behavior.
+
+RTVM with HV Emulated Device
+****************************
+
+ACRN uses hypervisor emulated virtual UART (vUART) devices for inter-VM synchronization
+such as logging output, or command send/receive. The vUART can only works in polling mode
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
such as logging output, or command send/receive. Currently, the vUART
only works in polling mode, but may be extended to support interrupt
mode in a future release.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+too at moment, and may be extended to support interrupt mode in future(TODO). In the
+meantime, the RT workload, using vUART, shall reserve margin of CPU cycles to accommodate
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
meantime, for better RT behavior, the RT workload using the vUART shall reserve a margin of CPU cycles to accommodate
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+the additional latency, introduced by the VM-Exit to the vUART I/O registers(~2000--3000
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
the additional latency introduced by the VM-Exit to the vUART I/O registers (~2000--3000
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+cycles per register access), for better RT behavior.
+
+DM emulated device (Except PMD) – DO NOT use
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: Change to:
```
DM emulated device (Except PMD)
*******************************
We recommend not using DM-emulated devices in a RTVM.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst (hunk 1 above)
Comment (JasonChenCJ) [reply doc/tutorials/rtvm_workload_design_guideline.rst]: @dbkinder , thanks for the comments, please review the new commit

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 14]
```diff
@@ -0,0 +1,44 @@
+.. _rtvm_workload_guideline:
+
+Guidelines to the workload design in an RTVM
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: Could we rephrase this as "Guidelines for designing an application to run in an RTVM"?

@dbkinder thoughts?

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+############################################
+
+An RTOS developer must be aware of the differences between running applications on a native
+platform and a real time virtual machine (RTVM), especially issues impacting real time
+performance. For example, a real time thread should avoid using VM-Exit. If VM-Exit is used,
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
performance. For example, a real time thread should avoid triggering any ``vmexit``. If ``vmexit`` are triggered,
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+the developer must account for an additional margin of CPU cycles for the added run time cost.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
the developer must account for an additional margin of CPU cycles for the incremental runtime.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+This document provides some workload design guidelines when using an RTVM within the ACRN hypervisor.
+
+Run RTVM with dedicated resources/devices
+*****************************************
+
+For best practice, ACRN allocates dedicated CPU, memory resources, and cache resources (with
+Cache Allocations Technology, i.e. CAT) to RTVMs. For best real time performance of I/O devices,
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
Cache Allocation Technology, aka CAT) to RTVMs. For best real time performance of I/O devices,
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+platform and a real time virtual machine (RTVM), especially issues impacting real time
+performance. For example, a real time thread should avoid using VM-Exit. If VM-Exit is used,
+the developer must account for an additional margin of CPU cycles for the added run time cost.
+
+This document provides some workload design guidelines when using an RTVM within the ACRN hypervisor.
+
+Run RTVM with dedicated resources/devices
+*****************************************
+
+For best practice, ACRN allocates dedicated CPU, memory resources, and cache resources (with
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
For best practice, ACRN allocates dedicated CPU, memory resources, and cache resources (using Intel
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+Cache Allocations Technology, i.e. CAT) to RTVMs. For best real time performance of I/O devices,
+we recommend using dedicated (pass-thru) PCIe devices to avoid VM-Exit at run time.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
we recommend using dedicated (pass-thru) PCIe devices to avoid ``vmexit`` at runtime.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+
+NOTE: The configuration space for pass-thru PCI devices is still emulated and the access will
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
.. note::
   The configuration space for pass-thru PCI devices is still emulated and the access will
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+trigger a VM-Exit.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
trigger a ``vmexit``.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+RTVM with virtio PMD (Polling Mode Driver) for I/O sharing
+**********************************************************
+
+If the RTVM must use shared devices, we recommend using PMD drivers that can eliminate the
+unpredictable latency caused by guest I/O trap-and-emulate access. The RTVM workload must be
+aware that the packets in the PMD driver may arrive or be sent later than expected. The workload
+in RTVM must be aware of for better RT behavior.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: I don't understand this last sentence, I suspect there is a word missing somewhere, or punctuation.

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+RTVM with HV Emulated Device
+****************************
+
+ACRN uses hypervisor emulated virtual UART (vUART) devices for inter-VM synchronization such as
+logging output, or command send/receive.  Currently, the vUART only works in polling mode, but
+may be extended to support interrupt mode in a future release. In the meantime, for better RT
+behavior, the RT workload using the vUART shall reserve a margin of CPU cycles to accommodate
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: @dbkinder shouldn't that be "to accomodate **for**"

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+the additional latency introduced by the VM-Exit to the vUART I/O registers(~2000--3000 cycles
+per register access), for better RT behavior.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: "for better RT behavior" is repeated twice in here.

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+DM emulated device (Except PMD)
+*******************************
+
+We recommend not using DM-emulated devices in a RTVM.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
We recommend to **not** use DM-emulated devices in an RTVM.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst (hunk 21 above)
Comment (dongyaozu) [reply doc/tutorials/rtvm_workload_design_guideline.rst]: SDM uses "VM-Exit", so our doc is better to stay with

**Comment by gvancuts** [comment]
SDM uses "VM-Exit", so our doc is better to stay with

Good point, agreed

**Comment by JasonChenCJ** [comment]
@gvancuts @dbkinder @dongyaozu, please help review again for the latest update based on your comments.

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+is triggered, the developer must account for an additional margin of CPU cycles for the
+incremental runtime.
+
+This document provides some workload design guidelines when using an RTVM within the ACRN hypervisor.
+
+Run RTVM with dedicated resources/devices
+*****************************************
+
+For best practice, ACRN allocates dedicated CPU, memory resources, and cache resources (using Intel
+Cache Allocations Technology, aka CAT) to RTVMs. For best real time performance of I/O devices,
```
Comment (gvancuts) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
Cache Allocation Technology, aka CAT) to RTVMs. For best real time performance of I/O devices,
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+   The configuration space for pass-thru PCI devices is still emulated and the access will
+   trigger a VM-Exit.
+
+RTVM with virtio PMD (Polling Mode Driver) for I/O sharing
+**********************************************************
+
+If the RTVM must use shared devices, we recommend using PMD drivers that can eliminate the
+unpredictable latency caused by guest I/O trap-and-emulate access. The RTVM workload must be
+aware that the packets in the PMD driver may arrive or be sent later than expected. The workload
+in RTVM must be aware of that for better RT behavior.
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]: We can delete this last sentence: ``The workload in ... for better RT behavior.``

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+
+RTVM with HV Emulated Device
+****************************
+
+ACRN uses hypervisor emulated virtual UART (vUART) devices for inter-VM synchronization such as
+logging output, or command send/receive.  Currently, the vUART only works in polling mode, but
+may be extended to support interrupt mode in a future release. In the meantime, for better RT
+behavior, the RT workload using the vUART shall reserve a margin of CPU cycles to accommodate
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
behavior, the RT application using the vUART should reserve a margin of CPU cycles to accommodate
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+for the additional latency introduced by the VM-Exit to the vUART I/O registers(~2000--3000 cycles
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
for the additional latency introduced by the VM-Exit to the vUART I/O registers (~2000-3000 cycles
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+per register access).
+
+DM emulated device (Except PMD)
+*******************************
+
+We recommend to **not** use DM-emulated devices in a RTVM.
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
We recommend **not** using DM-emulated devices in a RTVM.
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+
+Run RTVM with dedicated resources/devices
+*****************************************
+
+For best practice, ACRN allocates dedicated CPU, memory resources, and cache resources (using Intel
+Cache Allocation Technology, aka CAT) to RTVMs. For best real time performance of I/O devices,
+we recommend using dedicated (pass-thru) PCIe devices to avoid VM-Exit at run time.
+
+.. note::
+   The configuration space for pass-thru PCI devices is still emulated and the accessing it will
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
   The configuration space for pass-thru PCI devices is still emulated and accessing it will
```

File: doc/tutorials/rtvm_workload_design_guideline.rst [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+aware that the packets in the PMD driver may arrive or be sent later than expected.
+
+RTVM with HV Emulated Device
+****************************
+
+ACRN uses hypervisor emulated virtual UART (vUART) devices for inter-VM synchronization such as
+logging output, or command send/receive.  Currently, the vUART only works in polling mode, but
+may be extended to support interrupt mode in a future release. In the meantime, for better RT
+behavior, the RT application using the vUART shall reserve a margin of CPU cycles to accommodate
+for the additional latency introduced by the VM-Exit to the vUART I/O registers (~2000--3000 cycles
```
Comment (dbkinder) [inline doc/tutorials/rtvm_workload_design_guideline.rst]:
```suggestion
for the additional latency introduced by the VM-Exit to the vUART I/O registers (~2000-3000 cycles
```

**Comment by dbkinder** [comment]
I'll merge this and fix in a followup PR
